### PR TITLE
[integration-tests] extract daemon to its own package

### PR DESCRIPTION
**Pull request by vdemeester**
This extracts the integration-cli `Daemon` (and related) into its own package. This is more or less just an extract with a few fixes. It was a little bit trickier than I though at first as there is a lot a global variables used in the `integration-cli` code and loads of non exported fields used.

As it's a big PR already, I added comment to myself for the next step/refacto/enhancements. I tried to export as less as I could and did a few fixes on ignored errors :angel:.

The end goal is :
* Start the daemon from Go code and thus not requiring shell script to do it (a plain `go test` would do it :angel:).
* Start a daemon for all the suite.

One of the next step is to be able to use `Daemon` struct to start windows deamons :angel:.

/cc @thaJeztah @cpuguy83 @dnephin @icecrime @aaronlehmann @AkihiroSuda 

Some of the comments on the `daemon.Daemon` and `daemon.Swarm` are really bad because I did that quickly and mainly to pass the `golint` validation :sweat_smile:.

🐸

*hopefully this will be green :sweat_smile:*

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

### CODE REVIEW

File: integration-cli/docker_cli_swarm_test.go [hunk 1]
```diff
@@ -96,21 +97,23 @@ func (s *DockerSwarmSuite) TestSwarmInitUnspecifiedAdvertiseAddr(c *check.C) {
...
-	info, err := d.info()
+	info, err := d.DaemonInfo()
 	c.Assert(err, checker.IsNil)
 	c.Assert(info.LocalNodeState, checker.Equals, swarm.LocalNodeStateActive)
 	c.Assert(d.Stop(), checker.IsNil)
 
 	// start a daemon with --cluster-store and --cluster-advertise
 	err = d.Start("--cluster-store=consul://consuladdr:consulport/some/path", "--cluster-advertise=1.1.1.1:2375")
 	c.Assert(err, checker.NotNil)
-	content, _ := ioutil.ReadFile(d.logFile.Name())
```
Comment (vdemeester) [inline integration-cli/docker_cli_swarm_test.go]: Same below and in other places

File: integration-cli/daemon/daemon.go [hunk 2]
```diff
@@ -0,0 +1,831 @@
...
+
+	if err := d.cmd.Start(); err != nil {
+		return errors.Errorf("[%s] could not start daemon container: %v", d.id, err)
+	}
+
+	wait := make(chan error)
+
+	go func() {
+		wait <- d.cmd.Wait()
+		fmt.Printf("[%s] exiting daemon\n", d.id)
```
Comment (vdemeester) [inline integration-cli/daemon/daemon.go]: Those should probably be something else than `fmt.Printf` though :sweat:

File: integration-cli/daemon/daemon.go [hunk 3]
```diff
@@ -0,0 +1,831 @@
...
+	basesizeBytes := int64(basesizeFloat) * (1024 * 1024 * 1024)
+	return basesizeBytes
+}
+
+// Cmd will execute a docker CLI command against this Daemon.
+// Example: d.Cmd("version") will run docker -H unix://path/to/unix.sock version
+func (d *Daemon) Cmd(args ...string) (string, error) {
+	b, err := d.Command(args...).CombinedOutput()
+	return string(b), err
+}
```
Comment (dnephin) [inline integration-cli/daemon/daemon.go]: I really want to change this to use the interface from `pkg/integration/cmd/command.go`. Maybe I'll look at doing that once this is merged.

**Review by dnephin — APPROVED** [review]
Awesome change. Big +1

LGTM

File: integration-cli/daemon/daemon.go (hunk 3 above)
Comment (vdemeester) [inline integration-cli/daemon/daemon.go]: Oh dont worry to much, I am kinda already on it 😇

**Comment by AkihiroSuda** [comment]
windows failure seems unrelated
```
00:28:55.186 ----------------------------------------------------------------------
00:28:55.187 FAIL: docker_cli_ps_test.go:20: DockerSuite.TestPsListContainersBase
00:28:55.187 
00:28:55.187 docker_cli_ps_test.go:24:
00:28:55.187     out, _ = runSleepingContainer(c, "-d")
00:28:55.187 docker_utils.go:457:
00:28:55.187     c.Assert(result, icmd.Matches, icmd.Success)
00:28:55.187 ... result *cmd.Result = &cmd.Result{Cmd:(*exec.Cmd)(0xc042075600), ExitCode:125, Error:(*exec.ExitError)(0xc042fd0000), Timeout:false, outBuffer:(*cmd.lockedBuffer)(0xc042166bd0), errBuffer:(*cmd.lockedBuffer)(0xc042166c60)} ("\nCommand: d:\\CI\\CI-112d248ef\\binary\\docker.exe run -d -d busybox sleep 240\nExitCode: 125, Error: exit status 125\nStdout: e8ef2a6f6c509d23a8cac6e07d5ab8037968cbbc8a5d51665a6ee25933500811\n\nStderr: d:\\CI\\CI-112d248ef\\binary\\docker.exe: Error response from daemon: hcsshim::PrepareLayer failed in Win32: This operation returned because the timeout period expired. (0x5b4) layerId=e8ef2a6f6c509d23a8cac6e07d5ab8037968cbbc8a5d51665a6ee25933500811 flavour=1.\n\n")
00:28:55.190 ... expected cmd.Expected = cmd.Expected{ExitCode:0, Timeout:false, Error:"", Out:"", Err:""}
00:28:55.190 ... 
00:28:55.190 Command: d:\CI\CI-112d248ef\binary\docker.exe run -d -d busybox sleep 240
00:28:55.190 ExitCode: 125, Error: exit status 125
00:28:55.190 Stdout: e8ef2a6f6c509d23a8cac6e07d5ab8037968cbbc8a5d51665a6ee25933500811
00:28:55.190 
00:28:55.190 Stderr: d:\CI\CI-112d248ef\binary\docker.exe: Error response from daemon: hcsshim::PrepareLayer failed in Win32: This operation returned because the timeout period expired. (0x5b4) layerId=e8ef2a6f6c509d23a8cac6e07d5ab8037968cbbc8a5d51665a6ee25933500811 flavour=1.
00:28:55.191 
00:28:55.191 
00:28:55.191 Failures:
00:28:55.191 ExitCode was 125 expected 0
00:28:55.191 Expected no error
00:28:55.192 
00:28:55.192 
00:29:05.607 
00:29:05.607 ----------------------------------------------------------------------
```

**Comment by AkihiroSuda** [comment]
Can we silence a bunch of `[d657d57d2be4a] waiting for daemon to start` lines?
https://jenkins.dockerproject.org/job/Docker-PRs-experimental/28013/console

File: integration-cli/daemon/daemon.go [hunk 4]
```diff
@@ -0,0 +1,831 @@
...
+	return d.inspectFilter(name, fmt.Sprintf(".%s", field))
+}
+
+// FindContainerIP returns the ip of the specified container
+func (d *Daemon) FindContainerIP(id string) (string, error) {
+	out, err := d.Cmd("inspect", "--format='{{ .NetworkSettings.Networks.bridge.IPAddress }}'", id)
+	if err != nil {
+		return "", err
+	}
+	return strings.Trim(out, " \r\n'"), nil
```
Comment (AkihiroSuda) [inline integration-cli/daemon/daemon.go]: `strings.TrimSpace`?

File: integration-cli/daemon/daemon.go [hunk 5]
```diff
@@ -0,0 +1,831 @@
...
+		}
+
+		out := strings.TrimSpace(result.Stdout())
+		if out == expected {
+			break
+		}
+
+		select {
+		case <-after:
+			return errors.Errorf("condition \"%q == %q\" not true in time", out, expected)
```
Comment (AkihiroSuda) [inline integration-cli/daemon/daemon.go]: Can we put the timeout value to this error string?

**Review by AkihiroSuda — APPROVED** [review]
LGTM (but needs rebase)

**Comment by vdemeester** [comment]
Rebased :angel:

**Comment by aaronlehmann** [comment]
I suggest backporting this to 1.13. There is still a lot of code getting cherry-picked, and it will be difficult with changes like this in the integration test framework. Ideally I would have preferred that this wait until 1.13 is released.

**Pull request merged**